src/ExponentialBackoff.php

```diff
@@ -0,0 +1,176 @@
+    }
+
+    /**
+     * Get the minimum wait time
+     *
+     * @return int
+     */
+    public function getMin(): int
+    {
+        return $this->min;
+    }
+
+    /**
+     * Set the minimum wait time
+     *
+     * @param int $min
+     *
+     * @return $this
+     */
+    public function setMin(int $min): self
+    {
+        if ($min <= 0) {
+            $min = 100; // Default minimum wait time 100 ms
+        }
+
+        $this->min = $min;
+
+        return $this;
+    }
+
+    /**
+     * Get the maximum wait time
+     *
+     * @return int
+     */
+    public function getMax(): int
+    {
+        return $this->max;
+    }
+
+    /**
+     * Set the maximum wait time
+     *
+     * @param int $max
+     *
+     * @return $this
+     * @throws LogicException When the configured minimum wait time is greater than the maximum wait time
+     */
+    public function setMax(int $max): self
+    {
+        if ($max <= 0) {
+            $max = 10000; // Default max wait time 10 seconds
+        }
+
+        $this->max = $max;
+
+        if ($this->min > $this->max) {
+            throw new LogicException('Max must be larger than min');
+        }
+
+        return $this;
+    }
+
+    /**
+     * Get the configured number of retries
+     *
+     * @return int
+     */
+    public function getRetries(): int
+    {
+        return $this->retries;
+    }
+
+    /**
+     * Set number of retries to be used
+     *
+     * @param int $retries
+     *
+     * @return $this
+     */
+    public function setRetries(int $retries): self
+    {
+        $this->retries = $retries;
+
+        return $this;
+    }
+
+    /**
+     * Get a new wait time for the given attempt
+     *
+     * If the given attempt is the initial one, the min wait time is used. For all subsequent requests,
+     * the previous wait time is simply multiplied by 2.
+     *
+     * @param int $attempt
+     *
+     * @return int
+     */
+    public function getWaitTime(int $attempt): int
+    {
+        if ($attempt === 0) {
+            $this->previousWaitTime = null;
+        }
+
+        if ($this->previousWaitTime >= $this->max) {
+            return $this->max;
+        }
+
+        $next = min(! $this->previousWaitTime ? $this->min : $this->previousWaitTime * 2, $this->max);
+        $this->previousWaitTime = $next;
+
+        return $next;
+    }
+
+    /**
+     * Execute and retry the given callback
+     *
+     * @param callable(?Exception $err): mixed $callback The callback to be retried
+     *
+     * @return mixed
+     * @throws Exception When the given callback rethrows an exception that can't be retried or max retries is reached
+     */
+    public function retry(callable $callback)
+    {
+        $attempt = 0;
+        $previousErr = null;
+
+        do {
+            try {
+                return $callback($previousErr);
+            } catch (Exception $err) {
+                if ($attempt >= $this->getRetries() || $err === $previousErr) {
```

Comment:
Why is the logic for deciding whether something is retryable not part of this code, but must be executed in each individual callback? With the current implementation, the decision whether something is retryable is made after sleeping (the next time the callback is called), and the first attempt cannot stop the retryable logic (please add a test). And requires duplicate code in each individual callback.